core/src/test/java/feast/core/grpc/CoreServiceImplTest.java

```diff
@@ -0,0 +1,135 @@
+import feast.core.CoreServiceProto.CoreServiceTypes.GetUploadUrlRequest.FileType;
+import feast.core.CoreServiceProto.CoreServiceTypes.GetUploadUrlResponse;
+import feast.core.CoreServiceProto.CoreServiceTypes.GetUploadUrlResponse.HttpMethod;
+import feast.core.service.JobManagementService;
+import io.grpc.inprocess.InProcessChannelBuilder;
+import io.grpc.inprocess.InProcessServerBuilder;
+import io.grpc.testing.GrpcCleanupRule;
+import org.apache.commons.lang3.StringUtils;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.file.Files;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.junit.Assert.*;
+import static org.junit.Assume.assumeTrue;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class CoreServiceImplTest {
+  @Rule public final GrpcCleanupRule grpcCleanup = new GrpcCleanupRule();
+
+  @InjectMocks private CoreServiceImpl coreService = new CoreServiceImpl();
+
+  @Mock private JobManagementService jobManagementService;
+
+  @Test
+  public void getUploadUrl() throws IOException {
+    // Skip this test if GOOGLE_APPLICATION_CREDENTIALS is not set
+    // because this test requires Google service account key and connection to "actual"
+    // Google Cloud Storage service
+    assumeTrue(StringUtils.isNotEmpty(System.getenv("GOOGLE_APPLICATION_CREDENTIALS")));
+
+    // Mock the workspace output from jobManagementService to a bucket we can use for testing
+    String bucketName = "feast-templocation-kf-feast";
+    when(jobManagementService.getWorkspace()).thenReturn("gs://" + bucketName);
+
+    // Start a local grpc server
+    String serverName = InProcessServerBuilder.generateName();
+    grpcCleanup.register(
+        InProcessServerBuilder.forName(serverName)
+            .directExecutor()
+            .addService(coreService)
+            .build()
+            .start());
+    CoreServiceBlockingStub blockingStub =
+        CoreServiceGrpc.newBlockingStub(
+            grpcCleanup.register(
+                InProcessChannelBuilder.forName(serverName).directExecutor().build()));
+
+    GetUploadUrlResponse response =
+        blockingStub.getUploadUrl(
+            GetUploadUrlRequest.newBuilder().setFileType(FileType.CSV).build());
+
+    // Make sure the signed URL response allows upload using PUT method and the link is valid for
+    // the next 5 minutes
+    assertEquals(response.getHttpMethod(), HttpMethod.PUT);
+    assertTrue(response.getExpiration().getSeconds() > System.currentTimeMillis() / 1000);
+    assertTrue(
+        response.getExpiration().getSeconds() < (System.currentTimeMillis() + 6 * 60_000) / 1000);
```

Comment:
Where do these numbers come from?

Reply:
Updated, the value comes from the default duration for url validity (not configurable for now) in CoreServiceImpl